src/components/virtualRepeater/virtualRepeater.spec.js

```diff
@@ -0,0 +1,221 @@
+  });
+
+  it('should reposition and swap items on scroll (horizontal)', function() {
+    container.attr('md-horizontal', 'true');
+    createRepeater();
+    scope.items = createItems(NUM_ITEMS);
+    scope.$apply();
+    $$rAF.flush();
+
+    var repeated;
+
+    // Don't quite scroll past the first item.
+    scroller[0].scrollLeft = ITEM_SIZE - 1;
+    scroller.triggerHandler('scroll');
+    expect(offsetter.css('transform')).toBe('translateX(0px)');
+    repeated = getRepeated();
+    expect(repeated.length).toBe(HORIZONTAL_PX / ITEM_SIZE + VirtualRepeatController.NUM_EXTRA);
+    expect(repeated[0].textContent).toBe('s0s 0');
+
+    // Scroll past the first item.
+    // Expect that we now have the full set of extra items above and below.
+    scroller[0].scrollLeft = ITEM_SIZE;
+    scroller.triggerHandler('scroll');
+    expect(offsetter.css('transform')).toBe('translateX(0px)');
+    repeated = getRepeated();
+    expect(repeated.length)
+        .toBe(HORIZONTAL_PX / ITEM_SIZE + VirtualRepeatController.NUM_EXTRA + 1);
+    expect(repeated[0].textContent).toBe('s0s 0');
+
+    // Scroll past the fourth item.
+    // Expect that one new item is created.
+    scroller[0].scrollLeft = ITEM_SIZE * (VirtualRepeatController.NUM_EXTRA + 1);;
+    scroller.triggerHandler('scroll');
+    expect(offsetter.css('transform')).toBe('translateX(10px)');
+    repeated = getRepeated();
+    expect(repeated.length)
+        .toBe(HORIZONTAL_PX / ITEM_SIZE + VirtualRepeatController.NUM_EXTRA * 2);
+    expect(repeated[0].textContent).toBe('s2s 1');
+
+    // Scroll to the end.
+    // Expect the bottom extra items to be removed (pooled).
+    scroller[0].scrollLeft = 950;
+    scroller.triggerHandler('scroll');
+    expect(offsetter.css('transform')).toBe('translateX(920px)');
+    repeated = getRepeated();
+    expect(repeated.length).toBe(HORIZONTAL_PX / ITEM_SIZE + VirtualRepeatController.NUM_EXTRA);
+    expect(repeated[0].textContent).toBe('s184s 92');
+  });
+
+  it('should dirty-check only the swapped scope on scroll', function() {
+    createRepeater();
+    scope.items = createItems(NUM_ITEMS);
+    scope.$apply();
+    $$rAF.flush();
+    scroller[0].scrollTop = 100;
+    scroller.triggerHandler('scroll');
+
+    var scopes = Array.prototype.map.call(getRepeated(), function(elem) {
+      var s = angular.element(elem).scope();
+      spyOn(s, '$digest').and.callThrough();
+      return s;
+    });
+
+    // Scroll up by one.
+    // Expect only the last (index 15) scope to have $digested.
+    scroller[0].scrollTop = 90;
+    scroller.triggerHandler('scroll');
+    expect(scopes[15].$digest).toHaveBeenCalled();
+    expect(scopes[14].$digest).not.toHaveBeenCalled();
+
+    // Scroll down by two.
+    // Expect only the first scope to have $digested.
+    scroller[0].scrollTop = 110;
+    scroller.triggerHandler('scroll');
+    expect(scopes[0].$digest).toHaveBeenCalled();
+    expect(scopes[1].$digest).not.toHaveBeenCalled();
+  });
```

Comment:
Is it because the blocks/elements are pooled ?
So, the element at index 15 is removed from bottom, pooled and then used again at the top ?